tao_compiler/mlir/disc/transforms/disc_algebraic_simplifier.cc

```diff
@@ -90,6 +90,233 @@ struct MulOneTensorOp : public OpRewritePattern<mhlo::MulOp> {
+      std::vector<int64_t> newStridesVec;
+      std::vector<int64_t> newSliceResultShape;
+
+      int64_t flattenedStartIndice = 0, flattenedLimitIndice = 0;
+      int64_t dimAccum = 1;
+      for (int i = outterOperandShape.size() - 1; i >= 0; i--) {
+        flattenedStartIndice +=
+            dimAccum * (*(startIndicesAttr.begin() + i)).getSExtValue();
+        if (*(stridesAttr.begin() + i) != 1) {
+          return failure();
+        }
+        dimAccum *= outterOperandShape[i];
+      }
+      flattenedLimitIndice =
+          flattenedStartIndice +
+          op.getType().dyn_cast<RankedTensorType>().getNumElements() - 1;
+
+      dimAccum = 1;
+      for (int i = 0; i < innerOperandShape.size(); i++) {
+        dimAccum *= innerOperandShape[i];
+      }
+
+      size_t expected_ele_num =
+          op.getType().dyn_cast<RankedTensorType>().getNumElements();
+      size_t total_ele_num = 1;
+      bool shapeEnd = false;
+      for (int i = 0; i < innerOperandShape.size(); i++) {
+        dimAccum /= innerOperandShape[i];
+        newStartIndicesVec.push_back(flattenedStartIndice / dimAccum);
+        newLimitIndicesVec.push_back(flattenedLimitIndice / dimAccum + 1);
+        if (newLimitIndicesVec[i] <= newStartIndicesVec[i]) {
+          return failure();
+        }
+
+        newSliceResultShape.push_back(newLimitIndicesVec[i] -
+                                      newStartIndicesVec[i]);
+        newStridesVec.push_back(1);
+        total_ele_num *= (newLimitIndicesVec[i] - newStartIndicesVec[i]);
+        flattenedStartIndice %= dimAccum;
+        flattenedLimitIndice %= dimAccum;
+      }
+
+      if (total_ele_num != expected_ele_num) {
+        return failure();
+      }
+
+      auto newSliceResultType = mlir::RankedTensorType::get(
+          newSliceResultShape,
+          innerOperand.getType().dyn_cast<RankedTensorType>().getElementType());
+
+      auto newSliceOp = rewriter.create<mhlo::SliceOp>(
+          op.getLoc(), newSliceResultType, innerOperand,
+          rewriter.getI64TensorAttr(newStartIndicesVec),
+          rewriter.getI64TensorAttr(newLimitIndicesVec),
+          rewriter.getI64TensorAttr(newStridesVec));
+
+      auto newReshapeOp = rewriter.create<mhlo::ReshapeOp>(
+          op.getLoc(), op.getType(), newSliceOp);
+
+      rewriter.replaceOp(op, newReshapeOp.getResult());
+
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+struct ReshapeSimplifierPattern : public OpRewritePattern<mhlo::ReshapeOp> {
+  using OpRewritePattern<mhlo::ReshapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::ReshapeOp op,
+                                PatternRewriter& rewriter) const override {
+    if (matchPattern(op->getOperand(0), m_Op<mhlo::ReshapeOp>())) {
+      // Reshape(Reshape(A)) -> Reshape(A)
+      rewriter.replaceOp(op, op->getOperand(0));
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+struct OptimizationBarrierSimplifierPattern
```

Comment:
what's the purpose of this pattern rewriter?

Reply:
We will have 

```
B = All-Gather(A)
C = Reshape(B)
D = OptimizationBarrier(C)
E = Slice(D)
``
graph pattern lowered from `TorchAcc`, to prevent standalone `Reshape` Op, which will cause extra global memory read & write, we add this rule to change this graph into 

```
B = All-Gather(A)
C = OptimizationBarrier(B)
D = Reshape(C)
E = Slice(D)
``

@Yancey1989

Reply:
Thanks, after office discussion with @eedalong ,  it seems not a common pattern simplify rule, in the XLA FSDP implementation, the pattern `reshape->optimization_barrier->slice` is a work around on TPU, but in the NVGPU, we don't need that.
Maybe we can set `_shard_size_multiple=1`, and then this pattern would not exists . ref:  :https://github.com/AlibabaPAI/xla/blob/67edb354372a8e5cbce41f20ecfca68b328635c6/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L1435 .
cc  @anw90